components/layout_2020/query.rs

```diff
@@ -509,8 +514,415 @@ fn is_eligible_parent(fragment: &BoxFragment) -> bool {
+                // property.
+                let trim_beginning_white_space = !preserve_whitespace &&
+                    (state.may_start_with_whitespace || display == Display::InlineBlock);
+                let with_white_space_rules_applied: WhitespaceCollapse<_> = WhitespaceCollapse::new(
+                    text_content.chars(),
+                    white_space_collapse,
+                    trim_beginning_white_space,
+                );
+
+                // Step 4: If node is a Text node, then for each CSS text box produced by node, in
+                // content order, compute the text of the box after application of the CSS
+                // 'white-space' processing rules and 'text-transform' rules, set items to the list
+                // of the resulting strings, and return items. The CSS 'white-space' processing
+                // rules are slightly modified: collapsible spaces at the end of lines are always
+                // collapsed, but they are only removed if the line is the last line of the block,
+                // or it ends with a br element. Soft hyphens should be preserved.
+                let transformed_text = TextTransformation::new(
+                    with_white_space_rules_applied,
+                    style.clone_text_transform().case(),
+                )
+                .collect::<String>();
+
+                let is_preformatted_element =
+                    white_space_collapse == WhiteSpaceCollapseValue::Preserve;
+
+                let is_final_character_whitespace = transformed_text
+                    .chars()
+                    .next_back()
+                    .filter(|c| c.is_ascii_whitespace())
+                    .is_some();
+
+                let is_first_character_whitespace = transformed_text
+                    .chars()
+                    .next()
+                    .filter(|c| c.is_ascii_whitespace())
+                    .is_some();
+
+                // By truncating trailing white space and then adding it back in once we
+                // encounter another text node we can ensure no trailing white space for
+                // normal text without having to look ahead
+                if state.did_truncate_trailing_white_space && !is_first_character_whitespace {
+                    items.push(InnerOrOuterTextItem::Text(String::from(" ")));
+                };
+
+                if transformed_text.len() > 0 {
+                    // Here we decide whether to keep or truncate the final white
+                    // space character, if there is one.
+                    if is_final_character_whitespace && !is_preformatted_element {
+                        state.may_start_with_whitespace = false;
+                        state.did_truncate_trailing_white_space = true;
+                        let truncated_text = transformed_text
+                            .chars()
+                            .take(transformed_text.len() - 1)
+                            .collect();
```

Comment:
[`String::pop`](https://doc.rust-lang.org/std/string/struct.String.html#method.pop) also removes the last character, but without (necessarily) allocating a new string. I think this could be used here?

Reply:
Nice, works great